wdl-lint/src/fix.rs

```diff
@@ -0,0 +1,272 @@
+/// A replacement to be applied to a String.
+pub struct Replacement {
+    /// The start position of the replacement.
+    start: usize,
+    /// The end position of the replacement.
+    end: usize,
+    /// where to insert the replacement.
+    insertion_point: InsertionPoint,
+    /// Value to be inserted.
+    value: String,
+    /// Precedence for replacement. Higher precedences are applied first.
+    precedence: usize,
+}
+
+#[allow(unused)]
+impl Replacement {
+    /// Create a new `Replacement`.
+    pub fn new(
+        start: usize,
+        end: usize,
+        insertion_point: InsertionPoint,
+        value: String,
+        precedence: usize,
+    ) -> Self {
+        Replacement {
+            start,
+            end,
+            insertion_point,
+            value,
+            precedence,
+        }
+    }
+
+    /// The start position of the replacement.
+    pub fn start(&self) -> usize {
+        self.start
+    }
+
+    /// The end position of the replacement.
+    pub fn end(&self) -> usize {
+        self.end
+    }
+
+    /// where to insert the replacement.
+    pub fn insertion_point(&self) -> InsertionPoint {
+        self.insertion_point
+    }
+
+    /// Value to be inserted.
+    pub fn value(&self) -> &str {
+        self.value.as_ref()
+    }
+
+    /// Precedence for replacement. Higher precedences are applied first.
+    pub fn precedence(&self) -> usize {
+        self.precedence
+    }
+}
+
+// Adapted from ShellCheck's [Fixer](https://github.com/koalaman/shellcheck/blob/master/src/ShellCheck/Fixer.hs)
+/// Apply a series of `Replacement`s to a String.
+///
+/// Internally uses a [Fenwick Tree](https://en.wikipedia.org/wiki/Fenwick_tree)
+/// which is updated as replacements are applied. This allows multiple
+/// replacements referencing only the original input. Although the canonical
+/// Fenwick tree is 1-indexed this implementation is 0-indexed, so replacement
+/// positions must be directly equivalent to string indices.
+///
+/// The length of the tree is initialized to be 1 longer
+/// than the length of the initial value. This is because ftree provides
+/// no API for accessing the value of the final position, and the prefix sum
+/// only provides the cumulative sum < index. The extra index makes it possible
+/// to calculate the sum of the entire tree, which is necessary to enable
+/// slices of the new value beyond the original end position.
+/// Attempting to apply a replacement at this position will panic.
+#[derive(Clone, Debug)]
+pub struct Fixer {
+    /// The string to be modified.
+    value: String,
+    /// A Fenwick tree for tracking modifications.
+    tree: FenwickTree<i32>,
+}
+
+#[allow(unused)]
+impl Fixer {
+    /// Create a new Fixer from a String.
+    pub fn new(value: String) -> Self {
+        Fixer {
+            tree: FenwickTree::from_iter(vec![0; value.len() + 1]),
+            value,
+        }
+    }
+
+    /// Apply a [`Replacement`] to the value contained in the Fixer.
```

Comment:
Can we add a `# Panic` section mentioning that this will panic if the replacement is out of bounds?